java/ql/lib/semmle/code/java/dataflow/internal/DataFlowImpl.qll

```diff
@@ -147,6 +147,12 @@ abstract class Configuration extends string {
    */
   FlowFeature getAFeature() { none() }
 
+  /** Holds if sources should be grouped in the result of `hasFlowPath`. */
+  predicate sourceGrouping(Node source, string sourceGroup) { none() }
+
+  /** Holds if sinks should be grouped in the result of `hasFlowPath`. */
+  predicate sinkGrouping(Node sink, string sinkGroup) { none() }
+
```

Comment:
So just to make sure: If `conf.sourceGrouping(mySource1, "foo")` and `conf.sourceGrouping(mySource2, "foo")` holds, then any `PathNode` whose `getNode()` is `mySource1` or `mySource2` will be assigned the same `PathNodeSourceGroup`. Is that correct?

Is a given node allowed to be grouped into multiple `sourceGroups` (and similarly for sinkGroup)?

Reply:
Yes. Indeed there is a certain degree of flexibility here. The expectation is that you'll assign each sink to a single sink group, but assigning any combination of 0, 1, or 2+ groups to a sink is also going to work, but I'm not sure that the 2+ case is going to be useful. Think of this as augmenting the flow graph with some additional sink nodes (the groups) and adding some edges from the original sinks to those new nodes, and then have the flow be reported targeting those new sink nodes, if possible. Grouping several sinks will then allow you to reduce the total number of alerts. If you add multiple groups for a given sink, then you're effectively inflating the number of results, which I assume will be counterproductive.

Reply:
Thanks for the intuition!